pkg/kubelet/cadvisor/cadvisor_linux.go

```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
 	return cadvisorClient, nil
 }
 
+// New takes a memory storage and returns a new manager.
+func newCadvisor(memoryCache *memory.InMemoryCache, sysfs sysfs.SysFs, maxHousekeepingInterval time.Duration, allowDynamicHousekeeping bool, includedMetricsSet container.MetricSet, collectorHttpClient *http.Client, rawContainerCgroupPathPrefixWhiteList []string) (manager.Manager, error) {
+	if memoryCache == nil {
+		return nil, fmt.Errorf("manager requires memory storage")
+	}
+
+	// Detect the container we are running on.
+	selfContainer, err := cgroups.GetOwnCgroupPath("cpu")
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("cAdvisor running in container: %q", selfContainer)
+
+	var (
+		dockerStatus info.DockerStatus
+		rktPath      string
+	)
+	docker.SetTimeout(dockerClientTimeout)
+	// Try to connect to docker indefinitely on startup.
+	dockerStatus = retryDockerStatus()
+
+	crioClient, err := crio.Client()
+	if err != nil {
+		return nil, err
+	}
+	crioInfo, err := crioClient.Info()
+	if err != nil {
+		klog.V(5).Infof("CRI-O not connected: %v", err)
+	}
+
+	context := fs.Context{
+		Docker: fs.DockerContext{
+			Root:         docker.RootDir(),
+			Driver:       dockerStatus.Driver,
+			DriverStatus: dockerStatus.DriverStatus,
+		},
+		RktPath: rktPath,
+		Crio: fs.CrioContext{
+			Root: crioInfo.StorageRoot,
+		},
+	}
+	fsInfo, err := fs.NewFsInfo(context)
+	if err != nil {
+		return nil, err
+	}
+
+	// If cAdvisor was started with host's rootfs mounted, assume that its running
+	// in its own namespaces.
+	inHostNamespace := false
+	if _, err := os.Stat("/rootfs/proc"); os.IsNotExist(err) {
+		inHostNamespace = true
+	}
+
+	// Register for new subcontainers.
+	eventsChannel := make(chan watcher.ContainerEvent, 16)
+
+	machineInfo, err := machine.Info(sysfs, fsInfo, inHostNamespace)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(1).Infof("Machine: %+v", *machineInfo)
+
+	newManager := manager.New(
+		memoryCache,
+		fsInfo,
+		sysfs,
+		*machineInfo,
+		make([]chan error, 0, 2),
+		selfContainer,
+		inHostNamespace,
+		time.Now(),
+		maxHousekeepingInterval,
+		allowDynamicHousekeeping,
+		includedMetricsSet,
+		[]watcher.ContainerWatcher{},
+		eventsChannel,
+		collectorHttpClient,
+		&accelerators.NvidiaManager{},
+		rawContainerCgroupPathPrefixWhiteList)
+
+	versionInfo, err := newManager.GetVersionInfo()
+	if err != nil {
+		return nil, err
+	}
+	klog.V(1).Infof("Version: %+v", *versionInfo)
+	return newManager, nil
+}
+
+func retryDockerStatus() info.DockerStatus {
```

Comment:
There's a lot of code here copied from cadvisor. I'd prefer to see more of this exposed as methods for reuse, but I don't know that we need to block the PR on it.

Reply:
thanks!